index.js

```diff
@@ -20,6 +22,106 @@ app.get('/info', (req, res) => {
   res.send({ serverName: packageData.name, serverVersion: packageData.version });
 });
 
+// route for displaying all products
+app.get('/products/all', (req, res) => {
+
+  var jsonOutputObject = {
+    product_list_categorically : []
+  };
+
+  for(var i = 0; i < categoryData.categories.length; i++) {
+    
+    // JSON to be inserted in the main object
+    var jsonObjectToBeInserted = {
+      categoryId : categoryData.categories[i].id,
+      categoryName : categoryData.categories[i].categoryName,
+      products : []
+    }
+
+    for(var j = 0; j < productData.products.length; j++) {
+      if (productData.products[j].categoryId == categoryData.categories[i].id) {
+        jsonObjectToBeInserted.products.push(productData.products[j]);
+      }
+    }
+
+    jsonOutputObject.product_list_categorically.push(jsonObjectToBeInserted);
+  }
+
+  res.send(jsonOutputObject);
+});
+
+// Getting product info from its ID
+app.get('/product/:id', (req, res) => {
+
+  let userInput = req.params.id;
+  var productId;
+  var jsonOutputObject = null;
+
+  // Generating valid ID and returning error message when ID is invalid
+  if (userInput < 1) {
+    res.send('Invalid product ID');
+    return;
+  } else if (userInput >= 1 && userInput <= 9) {
+    productId = 'P00' + userInput;
+  } else {
+    productId = 'P0' + userInput;
+  }
+
+  for(var i=0; i<productData.products.length; i++) {
+    if(productData.products[i].id == productId) {
+      jsonOutputObject = productData.products[i];
+      break;
+    }
+  } 
+
+  if(jsonOutputObject == null) {
+    res.send('Product not found');
+  } else {
+    res.send(jsonOutputObject);
+  }
+});
+
+// Getting all the products under a category based on category's ID
+app.get('/category/:ctyId', (req, res) => {
+
+  let userInput = req.params.ctyId;
+  var categoryId;
+  var categoryNameVal;
+
+  var jsonOutputObject = {
+    category : categoryId,
+    categoryName : categoryNameVal,
+    products : []
+  };
+
+  // Generating valid ID and returning error message when ID is invalid
+  if (userInput < 1) {
+    res.send('Invalid category ID')
+    return;
+  } else if (userInput >= 1 && userInput <= 4) {
+    categoryId = 'cty0' + userInput;
+  } else {
+    res.send('Category not found!');
+    return;
+  }
+
+  // Getting category name
+  for(var i = 0; i < categoryData.categories.length; i++) {
+    if (categoryData.categories[i].id == categoryId) {
+      categoryNameVal = categoryData.categories[i].categoryName;
+      break;
+    }
+  }
+
+  for (var i = 0; i < productData.products.length; i++) {
+    if (productData.products[i].categoryId == categoryId) {
+      jsonOutputObject.products.push(productData.products[i]);
+    }
+  }
+
+  res.send(jsonOutputObject);
```

Comment:
Response needs to be array of products for that category

Reply:
Well, I've done exactly that, I've returned the array of products along with ctyId and categoryName as common (as repeating them in each array element would be redundant).

Hope this solves the issue :)